velox/common/base/BigintIdMap.h

```diff
@@ -0,0 +1,318 @@
+    auto resultArray = reinterpret_cast<int64_t*>(&resultVector);
+    uint16_t misses = matches ^ kAllSet;
+    while (misses) {
+      auto index = bits::getAndClearLastSetBit(misses);
+      int64_t byteOffset = 4 * (indexArray[index]);
+      if (UNLIKELY(byteOffset >= limit_)) {
+        byteOffset = 0;
+      }
+      for (;;) {
+        auto value = *reinterpret_cast<int64_t*>(table_ + byteOffset);
+        if (value == kEmptyMarker) {
+          resultArray[index] = kNotFound;
+          break;
+        }
+        if (value == dataArray[index]) {
+          resultArray[index] = *reinterpret_cast<uint32_t*>(
+              table_ + byteOffset + sizeof(int64_t));
+          break;
+        }
+        byteOffset += kEntrySize;
+        if (byteOffset >= limit_) {
+          byteOffset = 0;
+        }
+      }
+    }
+    return xsimd::load_unaligned(reinterpret_cast<int64_t*>(&resultVector));
+  }
+
+ private:
+  static constexpr int32_t kAllSet = bits::lowMask(xsimd::batch<int64_t>::size);
+  static constexpr int32_t kEntrySize = sizeof(int64_t) + sizeof(int32_t);
+  static constexpr int64_t kLow32 = (1L << 32) - 1;
+
+  // Number of bytes past end of last entry that may get read. The last id is
+  // accessed with a width of 8.
+  static constexpr int32_t kReadPadding = 4;
+
+  // Constants for hash calculation.
+  static constexpr uint64_t kMultLow = 1971049UL;
+  static constexpr uint64_t kMultHigh = 1470709UL;
+
+  // Allocates a new table.
+  void makeTable(int64_t capacity);
+
+  // Returns the pointer to the value of the 'i'th entry in 'table'.
+  int64_t* valuePtr(void* table, int32_t i) {
+    return reinterpret_cast<int64_t*>(
+        reinterpret_cast<char*>(table) + kEntrySize * i);
+  }
+
+  // Returns the pointer of the int32_t id for an entry.
+  int32_t* idPtr(int64_t* valuePtr) {
+    return reinterpret_cast<int32_t*>(valuePtr + 1);
+  }
+
+  // Rehashes 'this' to a size of 'newCapacity'.
+  void resize(int64_t newCapacity);
+
+  // Returns the hashed position of 'value' as a
+  // an index into an array of 12 byte entries. The function  is the same as
+  // indices()  for a single value. The difference is that indices returns
+  // distances in 4 byte words and this returns them i
+  int64_t indexOfEntry(int64_t value) {
+    uint32_t high = kMultHigh * (static_cast<uint64_t>(value) >> 32);
+    uint32_t low = kMultLow * static_cast<uint32_t>(value);
+    auto entry = ((high ^ low) & sizeMask_);
+    return entry;
+  }
+
+  // Sets the lanes of 'target' to the lanes of 'source' selected
+  // by 'mask'. Returns the result. If 'needsClear' is false, the
+  // set lanes of 'target' are assumed to be zero.
+
+  inline xsimd::batch<int64_t> setLanes(
+      xsimd::batch<int64_t> target,
+      xsimd::batch_bool<int64_t> mask,
+      xsimd::batch<int64_t> source,
+      bool needsClear = true) {
+    auto temp = mask;
+    auto maskAsInt64 = xsimd::load_unaligned(reinterpret_cast<int64_t*>(&temp));
```

Comment:
I will add a function in `SimdUtil.h` to make this portable

Reply:
@oerling xsimd already implemented this function for us, just use `xsimd::select(mask, source, target)`